### PR TITLE
fix: create dir for KOCACHE

### LINKED ISSUE #591: KOCACHE directory is not created if set
```
$ echo $KOCACHE
/Users/jasonhall/.ko
$ ls $KOCACHE
ls: /Users/jasonhall/.ko: No such file or directory
$ go run ./ build ./test/
... succeeds, takes a few seconds
$ go run ./ build ./test/
... succeeds, takes about the same amount of time
$ ls $KOCACHE
ls: /Users/jasonhall/.ko: No such file or directory
```

At this point, I would have expected for the `KOCACHE` directory to be created if I specified it, or at least complain in some way that it was specified but doesn't exist.

If I create it, I also need to `export` the env var to actually see the cache directory get populated.

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Fixes #591 

cc: @imjasonh

### CODE REVIEW

File: pkg/build/gobuild.go [hunk 1]
```diff
@@ -255,12 +255,17 @@ func build(ctx context.Context, ip string, dir string, platform v1.Platform, con
 		return "", fmt.Errorf("could not create env for %s: %w", ip, err)
 	}
 
-	tmpDir, err := ioutil.TempDir("", "ko")
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "ko")
```
Comment (imjasonh) [inline pkg/build/gobuild.go:258]: I'm not sure this change is necessary, at least not for the goal of this PR.

File: pkg/build/gobuild.go [hunk 2]
```diff
@@ -255,12 +255,17 @@ func build(ctx context.Context, ip string, dir string, platform v1.Platform, con
...
 	if err != nil {
 		return "", err
 	}
 
 	if dir := os.Getenv("KOCACHE"); dir != "" {
+		if _, err := os.Stat("dir"); os.IsNotExist(err) {
```
Comment (imjasonh) [inline pkg/build/gobuild.go:264]: This should be `dir`, not the string `"dir"`.

Comment (imjasonh) [inline pkg/build/gobuild.go:264]: We should check if the file exists and is a directory. If it's a file etc then KOCACHE won't work.

**Review by imjasonh — DISMISSED** [review]

**Comment by imjasonh** [comment]
We should think of some way to effectively test this behavior end to end. Maybe a test that sets `KOCACHE=/tmp/$(date +%s)` or something and checks that it exists after a build.

File: pkg/build/gobuild.go (hunk 2 above)
Comment (developer-guy) [reply pkg/build/gobuild.go]: thank you, good catch 🙈

File: pkg/build/gobuild.go (hunk 1 above)
Comment (developer-guy) [reply pkg/build/gobuild.go]: okay, I reverted that change.

File: pkg/build/gobuild.go (hunk 2 above)
Comment (developer-guy) [reply pkg/build/gobuild.go]: makes sense, I've updated the code according to your update, thanks a ton. 🙋🏻‍♂️

File: pkg/build/gobuild_test.go [hunk 3]
```diff
@@ -772,6 +774,42 @@ func TestGoBuild(t *testing.T) {
 	})
 }
 
+func TestGoBuildWithKOCACHE(t *testing.T) {
+	now := time.Now() // current local time
+	sec := now.Unix()
+	koCacheDir := "/tmp/"
```
Comment (jonjohnsonjr) [inline pkg/build/gobuild_test.go:780]: I think `/tmp/` might not always be writable. I'd set this to https://pkg.go.dev/testing#T.TempDir and ensure that the `filepath.Join(t.TempDir(), strconv.FormatInt(sec, 10))` directory exists.

Comment (developer-guy) [reply pkg/build/gobuild_test.go]: thank you, great advice, done. 🙋🏻‍♂️

**Comment by jonjohnsonjr** [comment]
Can you rebase this to pick up the new actions config?

**Comment by developer-guy** [comment]
> Can you rebase this to pick up the new actions config?

OFC 🙋🏻‍♂️

**Review by jonjohnsonjr — APPROVED** [review]

**Pull request merged**